### PR TITLE
fix update issue when using user assigned identity

### LINKED ISSUE #10406: r/kubernetes_cluster: api returns InvalidIdentityValues during update
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Terraform (and AzureRM Provider) Version

Terraform 0.14.5 (same behaviour with 0.14.4)
Azurerm 2.45.0 (same behaviour with 2.44.0)

### Affected Resource(s)

* `azurerm_kubernetes_cluster` when using identity.type="UserAssigned"

### Terraform Configuration Files

```hcl
resource "azurerm_user_assigned_identity" "aks" {
  resource_group_name = "myresourcegroup"
  location                         = "francecentral"
  name                             = "mycluster-identity"
}

resource "azurerm_kubernetes_cluster" "aks" {
  name     = "mycluster"
  location = "francecentral"
  
  resource_group_name = "myresourcegroup"
  node_resource_group  = "myresourcegroup-nodes"

  identity {
    type                                    = "UserAssigned"
    user_assigned_identity_id = azurerm_user_assigned_identity.aks.id
  }
  
  network_profile {
    network_plugin     = "kubenet"
    network_policy     = "calico"
    load_balancer_sku  = "Standard"
    outbound_type      = "userDefinedRouting"
    ...
  }

  private_cluster_enabled = false
  kubernetes_version        = "1.18.14"

  addon_profile {
    ...
  }

  default_node_pool {
    ...
    enable_auto_scaling = true
    min_count = 3
    max_count = 18
  }
```

### Debug Output

```
Error: updating Managed Kubernetes Cluster "mycluster" (Resource Group "myresourcegroup"): containerservice.ManagedClustersClient#CreateOrUpdate: Failure sending request: StatusCode=400 -- Original Error: Code="InvalidIdentityValues" Message="Invalid value for the identities '/subscriptions/xxxxxxxxx/resourceGroups/myresourcegroup/providers/Microsoft.ManagedIdentity/userAssignedIdentities/mycluster-identity'. The 'UserAssignedIdentities' property keys should only be empty json objects, null or the resource exisiting property."

  on aks.tf line 1, in resource "azurerm_kubernetes_cluster" "aks":
   1: resource "azurerm_kubernetes_cluster" "aks" {
```

### Expected Behaviour

The creation of azurerm_kubernetes_cluster worked without any problem. When I change a simple property (not related with user assigned identity), like default_node_pool.min_count I would expect the apply to not raise issues with UserAssignedIdentities

### Actual Behaviour

The creation of azurerm_kubernetes_cluster worked without any problem. When I change a simple property like default_node_pool.min_count or any other, I receive the error reported in Debug Output section.

### Steps to Reproduce

1. Create a cluster with user assigned identity (of course via terraform)
2. Change an azurerm_kubernetes_cluster property such as default_node_pool.min_count
3. `terraform plan`, no problem in the plan  
4. `terraform apply` and approve : error raised here

### References

Using UserAssigned identity, a feature released in 1.44.0 (https://github.com/terraform-providers/terraform-provider-azurerm/pull/8737/files)

### PR DESCRIPTION
**Pull request by njuCZ**
fix #10406

I have asked the aks service team, they say there is no such error in their codebase. The error may be thrown in ARM side or somewhere else.

according to the error message, if we set the value of `identity.UserAssignedIdentities` to empty, everything is ok. I think maybe somewhere the rule is strengthened, the computed value should not exist in the request body

additionally, I refactor the identity part and extract to a new file

image

### CODE REVIEW

**Comment by tombuildsstuff** [comment]
@njuCZ whilst this does workaround this API design flaw, ultimately this appears to be more of a modelling issue, for example:

https://github.com/Azure/azure-sdk-for-go/blob/52e3b873e1dc1eeb1d64e90754dbdc0d4851bab2/services/containerservice/mgmt/2020-12-01/containerservice/models.go#L1463-L1464
https://github.com/Azure/azure-sdk-for-go/blob/52e3b873e1dc1eeb1d64e90754dbdc0d4851bab2/services/containerservice/mgmt/2020-12-01/containerservice/models.go#L1479-L1485

Should likely be an array of User Assigned Identity ID's, or a list of nested blocks:

```
"identityIds": [
  {
    "id": "0000",
    # optional principal_id and tenant_id
  },
  {
    "id": "1111",
    # optional principal_id and tenant_id
  }
]
```

.. so can we file a Swagger bug about this, since this affects every resource using User Assigned Identities, so is a design flaw the API's which needs to be fixed there (although I can understand why it's been designed this way as a map, it's an odd design) - rather than working around in each consumer of the SDK/each service package in our case.

Thanks!

**Comment by clarenceb** [comment]
Hi, can this be merged and the design improvements moved to another issue?  It's causing pain for AKS users using the terraform azurerm provider and various workarounds needing to be used.

**Comment by ducas** [comment]
@tombuildsstuff I agree with @clarenceb. Not having this is preventing us from upgrading clusters using terraform, meaning we need to apply changes with the CLI and then sync state. Needless to say this is quite a pain...

The bug has been open for 7 weeks now so *something* needs to be done. I agree that the solution isn't ideal, but it's how the API works... :shrug:

**Comment by tombuildsstuff** [comment]
@clarenceb unfortunately this affects *every* resource in Azure that uses a User Assigned Identity - so whilst I can understand the desire to hot fix this in just for AKS, this fix would ultimately need to be applied to every resource.

From our side, we're happy to hot fix this API bug for AKS once a Swagger issue has been opened to track this, as [requested in this comment](https://github.com/terraform-providers/terraform-provider-azurerm/pull/10994#issuecomment-802786650) - but that's waiting on @njuCZ

**Comment by njuCZ** [comment]
@tombuildsstuff sorry for late response. I just tracked it internaly from email, still no further progress. This is the swagger issue: https://github.com/Azure/azure-rest-api-specs/issues/13631

**Comment by timja** [comment]
@tombuildsstuff issue is created, see above ^^, a hotfix would be appreciated.

Thanks!

**Review by katbyte — CHANGES_REQUESTED** [review]
Thanks for the PR @njuCZ but why are we refactoring so much in a bugfix pr? its hard to tell what is going on here and being changed. Could we please remove all refactoring and just include the fix for the issue? thanks

**Comment by njuCZ** [comment]
@katbyte done

File: azurerm/internal/services/containers/kubernetes_cluster_resource.go [hunk 1]
```diff
@@ -930,6 +930,8 @@ func resourceKubernetesClusterUpdate(d *schema.ResourceData, meta interface{}) e
 		return err
 	}
 
+	normalizeClusterUserAssignedIdentities(&existing)
```
Comment (katbyte) [inline azurerm/internal/services/containers/kubernetes_cluster_resource.go:933]: is there a reason to pull this out into a function could we inline it?

**Review by katbyte — APPROVED** [review]
Thanks @njuCZ - LGTM 👍

**Pull request merged**